From 7a554d4c9f86510bd8c10b0f12bc875949f2a4d7 Mon Sep 17 00:00:00 2001
From: Misael Lopez Cruz <misael.lopez@ti.com>
Date: Thu, 6 Jan 2011 01:54:45 -0600
Subject: ASoC: ABE DSP: Fix opp shadow variable for suspend/resume

ABE driver keeps track of last OPP used (25, 50, 100) in order to
avoid multiple OPP transitions. When the system goes to suspend/resume,
none of the supported OPP values is appropriate for that state since
VDD_IVA voltage remains in RET voltage level. It's required to set
ABE OPP again after suspend/resume scenario. During suspend() call,
ABE's opp shadow variable should be different to any of the valid OPPs
to allow OPP transition to happen in the next playback/capture operation
after suspend/resume.

Change-Id: Ia228469826928763e4c50e9f4d9b6ec2b4f04a18
Signed-off-by: Misael Lopez Cruz <misael.lopez@ti.com>
---
 sound/soc/omap/omap-abe-dsp.c |   12 +++++++++++-
 1 files changed, 11 insertions(+), 1 deletions(-)

Index: Kernel/sound/soc/omap/omap-abe-dsp.c
===================================================================
--- Kernel.orig/sound/soc/omap/omap-abe-dsp.c	2011-01-19 00:14:54.873323913 +0530
+++ Kernel/sound/soc/omap/omap-abe-dsp.c	2011-01-19 00:19:22.862073990 +0530
@@ -1657,8 +1657,12 @@
 			break;
 		}
 	} else if (abe->opp < opp) {
-		/* Increase OPP mode - no need of OPP25% */
+		/* Increase OPP mode */
 		switch (opp) {
+		case 25:
+			omap_device_set_rate(&pdev->dev, &pdev->dev, 49000000);
+			abe_set_opp_processing(ABE_OPP25);
+			break;
 		case 50:
 			omap_device_set_rate(&pdev->dev, &pdev->dev, 98000000);
 			abe_set_opp_processing(ABE_OPP50);
@@ -1926,6 +1930,12 @@
 
 	pm_runtime_put_sync(&pdev->dev);
 
+	/*
+	 * force setting OPP after suspend/resume to ensure
+	 * ABE freq/volt are set to proper values
+	 */
+	abe->opp = 0;
+
 	if (pdata->get_context_loss_count)
 		abe->loss_count = pdata->get_context_loss_count(dev);
 
